### PR TITLE
[BEAM-9679] Add Additional Outputs Lesson to Core Transform Go SDK katas

### PR DESCRIPTION
**Pull request by damondouglas**
This pull requests adds an Additional Outputs lesson to the Go SDK katas.  I would like to request the following reviewers:

(R: @lostluck )
(R: @henryken )

If accepted by both reviewers, please wait until the [Stepik course](https://stepik.org/course/70387) is updated before finally merging this PR.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark
--- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by lostluck** [comment]
I have not forgotten about this PR, I'm unfortunately a bit distrcted until later this afternoon. I should be able to get to it then. Thank you for your patience!

**Comment by damondouglas** [comment]
No problem at all.  Please take your time and thank you.

File: learning/katas/go/Core Transforms/Side Output/Side Output/pkg/task/task.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+	"github.com/apache/beam/sdks/go/pkg/beam"
+)
+
+func ApplyTransform(s beam.Scope, input beam.PCollection) (beam.PCollection, beam.PCollection) {
+	return beam.ParDo2(s, func(element int, numBelow100, numAbove100 func(int)) {
+		if element <= 100 {
+			numBelow100(element)
+			return
+		}
+		numAbove100(element)
```
Comment (henryken) [inline learning/katas/go/Core Transforms/Side Output/Side Output/pkg/task/task.go:28]: While this works too, maybe you would want to consider using `else` statement.

Comment (lostluck) [reply learning/katas/go/Core Transforms/Side Output/Side Output/pkg/task/task.go:28]: As written this is idiomatic Go.

File: learning/katas/go/Core Transforms/Side Output/Side Output/task.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+
+# Side Output
+
+While ParDo always produces a main output PCollection (as the return value from apply), you can 
+also have your ParDo produce any number of additional output PCollections. If you choose to have 
+multiple outputs, your ParDo returns all of the output PCollections (including the main output) 
+bundled together.
```
Comment (lostluck) [inline learning/katas/go/Core Transforms/Side Output/Side Output/task.md:24]: What's the "main output" ? Consider removing the paranthetical, I don't think it adds anything. 

DoFns in BeamGo don't have a "main" output, and we do not use that term in the GoDoc (which has problems, but this isn't one of them). They have the first, second, third etc outputs. DoFn return outputs only have two properties, if present, they're always the first one, and they require the DoFn to always return an element (or abort with an error).

File: learning/katas/go/Core Transforms/Side Output/Side Output/pkg/task/task.go (hunk 1 above)
Comment (damondouglas) [reply learning/katas/go/Core Transforms/Side Output/Side Output/pkg/task/task.go]: I will keep this unchanged then.  Thank you.

**Comment by damondouglas** [comment]
Thank you, Robert.  @henryken This has significantly changed since you agreed with it earlier aside from the `else` syntax.  I just wanted to make sure you were ok before I updated Stepik.  Of course any recommended changes are more than welcome from you.

File: learning/katas/go/Core Transforms/Additional Outputs/Additional Outputs/task.md [hunk 3]
```diff
@@ -16,12 +16,10 @@
   ~ limitations under the License.
   -->
 
-# Side Output
+# Additional Outputs
 
-While ParDo always produces a main output PCollection (as the return value from apply), you can
```
Comment (henryken) [inline learning/katas/go/Core Transforms/Additional Outputs/Additional Outputs/task.md]: This paragraph is actually from the Beam Programming Guide.
https://beam.apache.org/documentation/programming-guide/#additional-outputs
Probably should consider revising the guide as well if it's not accurate, e.g. main output.

FYI, historically it used to be called Side Output.

Comment (henryken) [reply learning/katas/go/Core Transforms/Additional Outputs/Additional Outputs/task.md]: I'll revise the name in other Katas as well

**Comment by henryken** [comment]
@damondouglas, I find the changes are okay. You may proceed.
I left a comment above.

**Comment by damondouglas** [comment]
@lostluck / @henryken I've updated [the stepik course](https://stepik.org/course/70387) and committed the modified `*-remote.yaml` files to this PR.  It is ready to merge.

**Comment by lostluck** [comment]
Thanks! Have a good weekend.

File: learning/katas/go/Core Transforms/Additional Outputs/Additional Outputs/task.md (hunk 3 above)
Comment (lostluck) [reply learning/katas/go/Core Transforms/Additional Outputs/Additional Outputs/task.md]: I remember. Side Output was a historical holdover from the original Dataflow API, and by the nature of how extra outputs are handled by the Java SDK. Thanks!

**Pull request merged**